parsl/executors/radical/executor.py

```diff
@@ -0,0 +1,388 @@
+                                 'flags': rp.DEFAULT_FLAGS}]
+            tds.append(td)
+
+        self.pmgr = rp.PilotManager(session=self.session)
+        self.tmgr = rp.TaskManager(session=self.session)
+
+        # submit pilot(s)
+        pilot = self.pmgr.submit_pilots(pd)
+        self.tmgr.submit_tasks(tds)
+
+        # FIXME: master tasks are never checked for, should add a task state
+        #        callback to do so.
+
+        pilot.stage_in({'source': ru.which('radical-pilot-hello.sh'),
+                        'target': 'radical-pilot-hello.sh',
+                        'action': rp.TRANSFER})
+
+        python_v = sys.version.split(' ')[0]
+        pilot.prepare_env(env_name='ve_raptor',
+                          env_spec={'type': self.pilot_env.get('type', 'virtualenv'),
+                                    'version': self.pilot_env.get('version', python_v),
+                                    'path': self.pilot_env.get('path', ''),
+                                    'pre_exec': self.pilot_env.get('pre_exec', []),
+                                    'setup': self.pilot_env.get('setup', [])})
+
+        self.tmgr.add_pilots(pilot)
+        self.tmgr.register_callback(self.task_state_cb)
+
+        # create a bulking thread to run the actual task submittion to RP in
+        # bulks
+        if self.bulk_mode:
+            self._max_bulk_size = 1024
+            self._max_bulk_time = 3        # seconds
+            self._min_bulk_time = 0.1      # seconds
+
+            self._bulk_queue = queue.Queue()
+            self._bulk_thread = mt.Thread(target=self._bulk_collector)
+
+            self._bulk_thread.daemon = True
+            self._bulk_thread.start()
+
+        return True
+
+    def unwrap(self, func, args):
+
+        task_type = ''
+
+        # Ignore the resource dict from Parsl
+        new_args = list(args)
+        new_args.pop(0)
+        args = tuple(new_args)
+
+        # remove the remote wrapper from parsl
+        while hasattr(func, '__wrapped__'):
+            func = func.__wrapped__
+
+        # identify the task type
+        try:
+            # bash and python might be partial wrapped
+            if isinstance(func, partial):
+                # @bash_app from parsl
+                try:
+                    task_type = inspect.getsource(func.args[0]).split('\n')[0]
+                    if BASH in task_type:
+                        task_type = BASH
+                        func = func.args[0]
+                    else:
+                        task_type = PYTHON
+
+                except Exception:
+                    logger.exception('unwrap failed')
+
+                    return func, args, task_type
+
+            # @python_app from parsl
+            else:
+                task_type = inspect.getsource(func).split('\n')[0]
+                if PYTHON in task_type:
+                    task_type = PYTHON
+                else:
+                    task_type = ''
```

Comment:
is this an error case?

Reply:
Yes, as sometimes we fail to extract the task type.